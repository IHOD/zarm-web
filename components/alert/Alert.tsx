import React, { Component } from 'react';
import classnames from 'classnames';
import Icon from '../icon';
import AlertProps from './PropsType';

class Alert extends Component<AlertProps, any> {
  static defaultProps = {
    prefixCls: 'ui-alert',
    type: 'info',
    title: '',
    description: null,
    width: 500,
    showIcon: false,
    closable: false,
    closeText: null,
    preventHide: false,
    onClick: () => {},
    onClose: () => {},
    afterClose: () => {},
  };

  domRef: HTMLDivElement | null;

  constructor(props) {
    super(props);
    this.state = {
      // 是否展示提示框
      visibility: true,
    };
  }

  componentDidMount() {
    if (this.domRef) {
      this.domRef.style.height = `${this.domRef.offsetHeight}px`;
    }
  }

  close = () => {
    const { afterClose } = this.props;
    window.setTimeout(() => {
      this.setState({ visibility: false });
      afterClose();
    }, 550);
    if (this.domRef) {
      this.domRef.style.height = '0px';
      this.domRef.style.paddingTop = '0px';
      this.domRef.style.paddingBottom = '0px';
      this.domRef.style.marginBottom = '0px';
    }
  };

  handleClose = () => {
    const { preventHide, onClick, onClose } = this.props;
    if (preventHide) {
      onClick({
        close: this.close.bind(this),
      });
    } else {
      this.close();
      onClose();
    }
  };

  render() {
    const {
      prefixCls,
      type,
      title,
      width,
      className,
      style,
      closable,
      closeText,
      description,
      showIcon,
    } = this.props;

    const { visibility, closeStart } = this.state;

    return (
      visibility && (
        <div
          ref={(domRef) => { this.domRef = domRef; }}
          style={{ width, ...style }}
          className={classnames(prefixCls, className, {
            [`${prefixCls}-${type}`]: Boolean(type),
            [`${prefixCls}-closable`]: closable,
            [`${prefixCls}-no-with-description`]: !description,
            [`${prefixCls}-sm-icon`]: showIcon && !description,
            [`${prefixCls}-lg-icon`]: showIcon && Boolean(description),
            [`${prefixCls}-close-animation`]: closeStart,
          })}
        >
          {showIcon && <AlertIcon {...this.props} />}
          {closable && <CloseIcon closable={closable} closeText={closeText} prefixCls={prefixCls} handleClose={this.handleClose} close={this.close} /> }
          <span className={`${prefixCls}-message`}>{title}</span>
          <span className={`${prefixCls}-description`}>{description}</span>
        </div>
      )
    );
  }
}

/**
 * 适配icon
 * @param props
 */
function AlertIcon(props) {
  const { description, type, icon } = props;
  // 无描述显示小图， 有描述显示大图
  // 优先选用用户传入的icon类型
  // 适配type
  if (icon && icon.isReactComponent) {
    return icon;
  }
  let iconType;
  switch (type) {
    case 'warning':
      iconType = 'warning-round-fill';
      break;
    case 'success':
      iconType = 'right-round-fill';
      break;
    case 'error':
      iconType = 'wrong-round-fill';
      break;
    case 'info':
    default:
      iconType = 'info-round-fill';
      break;
  }
  if (description === null) {
    return <Icon type={iconType} size="sm" />;
  }
  return <Icon type={iconType} size="lg" />;
}

/**
 * 适配关闭按钮
 * @param param
 */
function CloseIcon({ closable, closeText, prefixCls, handleClose, close }) {
  if (!closable) {
    return null;
  }

  if (!closeText) {
    return (
      <Icon className={`${prefixCls}-close-btn`} type="wrong" onClick={handleClose} />
    );
  }

  if (Object.is(closeText.constructor, String)) {
    return <button className={`${prefixCls}-custom-close-btn`} onClick={handleClose}>{ closeText }</button>;
  }

  if (Object.is(closeText.constructor, Function)) {
    const ReactNode = closeText;
    return <ReactNode close={close} />;
  }

  return closeText;
}

export default Alert;
