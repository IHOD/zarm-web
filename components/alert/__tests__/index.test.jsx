import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Alert from '../index';

describe('Alert', () => {
  it('renders basic normal Alert correctly', () => {
    const wrapper = mount(
      <Alert
        type="info"
        title="测试经典无描述，无icon，无关闭提示框"
      />,
      <Alert
        type="error"
        title="测试经典无描述，无icon，无关闭提示框"
      />,
      <Alert
        type="success"
        title="测试经典无描述，无icon，无关闭提示框"
      />,
      <Alert
        type="warning"
        title="测试经典无描述，无icon，无关闭提示框"
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders basic normal Alert correctly', () => {
    const wrapper = mount(
      <Alert showIcon title="显示icon-信息提示" type="info"/>,
      <Alert showIcon title="显示icon-成功" type="success" />,
      <Alert showIcon title="显示icon-警告" type="warning" />,
      <Alert showIcon title="显示icon-错误" type="error" />,
      <Alert showIcon title="显示icon-信息提示" description="显示大图标" type="info"/>,
      <Alert showIcon title="显示icon-成功" description="显示大图标" type="success" />,
      <Alert showIcon title="显示icon-警告" description="显示大图标"  type="warning" />,
      <Alert showIcon title="显示icon-错误" description="显示大图标" type="error" />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders basic normal Alert correctly', () => {
    const wrapper = mount(
      <Alert closable title="基础用法-信息提示" description="可关闭提示框" type="info"/>,
      <Alert
        closable
        title="基础用法-警告"
        type="warning"
        description="可关闭提示框"
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders basic normal Alert correctly', () => {
    function CustomText () {
      return (
        <span
          style={{
            fontSize: 12,
            color: '#909090',
            position: 'absolute',
            top: 8,
            right: 16,
            lineHeight: '22px',
            cursor: 'pointer'
          }}>不再提示</span>
      )
    }
    const wrapper = mount(
      <Alert
        closable
        title="自定义关闭按钮"
        type="info"
        closeText="查看详情"
      />,
      <Alert
        closable
        closeText={<CustomText />}
        title="自定义关闭按钮"
        type="warning"
        description="自定义文字"
      />,
      <Alert
        closable
        closeText={<Icon className="ui-alert-custom-close-btn" type="wrong-round-fill"/>}
        title="基础用法-错误"
        type="error"
        description="自定义关闭icon"
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
