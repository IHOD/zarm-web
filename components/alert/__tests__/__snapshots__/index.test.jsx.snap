// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders basic Alert correctly 1`] = `
<ForwardRef(forwardRef)
  message="这是一个警告框"
  onClose={[MockFunction]}
  theme="warning"
  visible={true}
>
  <LocaleReceiver
    forwardedRef={null}
    message="这是一个警告框"
    onClose={[MockFunction]}
    theme="warning"
    visible={true}
  >
    <AlertExtension
      className=""
      closable={true}
      hideIcon={false}
      locale={
        Object {
          "close": "关闭",
        }
      }
      localeCode="zh-cn"
      message="这是一个警告框"
      onClose={[MockFunction]}
      prefixCls="ui-alert"
      theme="warning"
      visible={true}
      width={270}
    >
      <Modal
        animationDuration={300}
        animationType="zoom"
        className=""
        isRadius={false}
        isRound={false}
        minWidth={270}
        onMaskClick={[Function]}
        prefixCls="ui-modal"
        visible={true}
        width={270}
      >
        <Portal
          containerInfo={
            <div>
              <div
                class="ui-modal fade-enter"
                style="animation-duration: 300ms; position: fixed;"
              >
                <div
                  class="ui-modal-wrapper"
                >
                  <div
                    class="ui-modal-dialog zoom-enter"
                    style="width: 270px; min-width: 270px; animation-duration: 300ms;"
                    tabindex="-1"
                  >
                    <div
                      class="ui-modal-body"
                      style="height: auto;"
                    >
                      <div
                        class="ui-alert"
                      >
                        <i
                          class="zw-icon zw-icon--warning-round zw-icon--warning zw-icon--md"
                          theme="warning"
                          type="warning-round"
                        >
                          <svg
                            fill="currentColor"
                            height="1em"
                            viewBox="0 0 32 32"
                            width="1em"
                          >
                            <use
                              xlink:href="#warning-round"
                            />
                          </svg>
                        </i>
                        <span>
                          这是一个警告框
                        </span>
                      </div>
                    </div>
                    <div
                      class="ui-modal-footer"
                    >
                      <button
                        class="zw-button zw-button--default zw-button--radius"
                        type="button"
                      >
                        <span>
                          关闭
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          }
        >
          <div
            className="ui-modal fade-enter"
            onClick={[Function]}
            style={
              Object {
                "animationDuration": "300ms",
                "position": "fixed",
              }
            }
          >
            <div
              className="ui-modal-wrapper"
            >
              <div
                className="ui-modal-dialog zoom-enter"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                style={
                  Object {
                    "animationDuration": "300ms",
                    "minWidth": 270,
                    "width": 270,
                  }
                }
                tabIndex={-1}
              >
                <ModalBody
                  height="auto"
                  prefixCls="ui-modal"
                  style={Object {}}
                >
                  <div
                    className="ui-modal-body"
                    style={
                      Object {
                        "height": "auto",
                      }
                    }
                  >
                    <div
                      className="ui-alert"
                    >
                      <Icon
                        prefixCls="zw-icon"
                        size="md"
                        theme="warning"
                        type="warning-round"
                        viewBox="0 0 32 32"
                      >
                        <Iconfont
                          className="zw-icon zw-icon--warning-round zw-icon--warning zw-icon--md"
                          size="md"
                          theme="warning"
                          type="warning-round"
                        >
                          <Icon
                            className="zw-icon zw-icon--warning-round zw-icon--warning zw-icon--md"
                            prefixCls="zw-icon"
                            size="md"
                            theme="warning"
                            type="warning-round"
                            viewBox="0 0 32 32"
                          >
                            <i
                              className="zw-icon zw-icon--warning-round zw-icon--warning zw-icon--md"
                              size="md"
                              theme="warning"
                              type="warning-round"
                            >
                              <svg
                                fill="currentColor"
                                height="1em"
                                viewBox="0 0 32 32"
                                width="1em"
                              >
                                <use
                                  xlinkHref="#warning-round"
                                />
                              </svg>
                            </i>
                          </Icon>
                        </Iconfont>
                      </Icon>
                      <span>
                        这是一个警告框
                      </span>
                    </div>
                  </div>
                </ModalBody>
                <ModalFooter
                  prefixCls="ui-modal"
                  style={Object {}}
                >
                  <div
                    className="ui-modal-footer"
                    style={Object {}}
                  >
                    <Button
                      ghost={false}
                      htmlType="button"
                      icon=""
                      loading={false}
                      onClick={[MockFunction]}
                      prefixCls="zw-button"
                      shape="radius"
                      size={null}
                      theme="default"
                    >
                      <button
                        className="zw-button zw-button--default zw-button--radius"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          关闭
                        </span>
                      </button>
                    </Button>
                  </div>
                </ModalFooter>
              </div>
            </div>
          </div>
        </Portal>
      </Modal>
    </AlertExtension>
  </LocaleReceiver>
</ForwardRef(forwardRef)>
`;

exports[`Alert renders basic normal Alert correctly 1`] = `
<ForwardRef(forwardRef)
  title="测试经典无描述，无icon，无关闭提示框"
  type="info"
>
  <LocaleReceiver
    forwardedRef={null}
    title="测试经典无描述，无icon，无关闭提示框"
    type="info"
  >
    <Alert
      afterClose={[Function]}
      closable={false}
      closeText={null}
      description={null}
      locale={
        Object {
          "close": "关闭",
        }
      }
      localeCode="zh-cn"
      onClick={[Function]}
      onClose={[Function]}
      prefixCls="ui-alert"
      preventHide={false}
      showIcon={false}
      title="测试经典无描述，无icon，无关闭提示框"
      type="info"
      width={500}
    >
      <div
        className="ui-alert ui-alert-info ui-alert-no-with-description"
        style={
          Object {
            "width": 500,
          }
        }
      >
        <span
          className="ui-alert-message"
        >
          测试经典无描述，无icon，无关闭提示框
        </span>
        <span
          className="ui-alert-description"
        />
      </div>
    </Alert>
  </LocaleReceiver>
</ForwardRef(forwardRef)>
`;

exports[`Alert renders basic normal Alert correctly 2`] = `
<ForwardRef(forwardRef)
  showIcon={true}
  title="显示icon-信息提示"
  type="info"
>
  <LocaleReceiver
    forwardedRef={null}
    showIcon={true}
    title="显示icon-信息提示"
    type="info"
  >
    <Alert
      afterClose={[Function]}
      closable={false}
      closeText={null}
      description={null}
      locale={
        Object {
          "close": "关闭",
        }
      }
      localeCode="zh-cn"
      onClick={[Function]}
      onClose={[Function]}
      prefixCls="ui-alert"
      preventHide={false}
      showIcon={true}
      title="显示icon-信息提示"
      type="info"
      width={500}
    >
      <div
        className="ui-alert ui-alert-info ui-alert-no-with-description ui-alert-sm-icon"
        style={
          Object {
            "width": 500,
          }
        }
      >
        <AlertIcon
          afterClose={[Function]}
          closable={false}
          closeText={null}
          description={null}
          locale={
            Object {
              "close": "关闭",
            }
          }
          localeCode="zh-cn"
          onClick={[Function]}
          onClose={[Function]}
          prefixCls="ui-alert"
          preventHide={false}
          showIcon={true}
          title="显示icon-信息提示"
          type="info"
          width={500}
        >
          <Icon
            prefixCls="zw-icon"
            size="sm"
            theme="default"
            type="info-round-fill"
            viewBox="0 0 32 32"
          >
            <Iconfont
              className="zw-icon zw-icon--info-round-fill zw-icon--default zw-icon--sm"
              size="sm"
              theme="default"
              type="info-round-fill"
            >
              <Icon
                className="zw-icon zw-icon--info-round-fill zw-icon--default zw-icon--sm"
                prefixCls="zw-icon"
                size="sm"
                theme="default"
                type="info-round-fill"
                viewBox="0 0 32 32"
              >
                <i
                  className="zw-icon zw-icon--info-round-fill zw-icon--default zw-icon--sm"
                  size="sm"
                  theme="default"
                  type="info-round-fill"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    viewBox="0 0 32 32"
                    width="1em"
                  >
                    <use
                      xlinkHref="#info-round-fill"
                    />
                  </svg>
                </i>
              </Icon>
            </Iconfont>
          </Icon>
        </AlertIcon>
        <span
          className="ui-alert-message"
        >
          显示icon-信息提示
        </span>
        <span
          className="ui-alert-description"
        />
      </div>
    </Alert>
  </LocaleReceiver>
</ForwardRef(forwardRef)>
`;

exports[`Alert renders basic normal Alert correctly 3`] = `
<ForwardRef(forwardRef)
  closable={true}
  description="可关闭提示框"
  title="基础用法-信息提示"
  type="info"
>
  <LocaleReceiver
    closable={true}
    description="可关闭提示框"
    forwardedRef={null}
    title="基础用法-信息提示"
    type="info"
  >
    <Alert
      afterClose={[Function]}
      closable={true}
      closeText={null}
      description="可关闭提示框"
      locale={
        Object {
          "close": "关闭",
        }
      }
      localeCode="zh-cn"
      onClick={[Function]}
      onClose={[Function]}
      prefixCls="ui-alert"
      preventHide={false}
      showIcon={false}
      title="基础用法-信息提示"
      type="info"
      width={500}
    >
      <div
        className="ui-alert ui-alert-info ui-alert-closable"
        style={
          Object {
            "width": 500,
          }
        }
      >
        <CloseIcon
          closable={true}
          close={[Function]}
          closeText={null}
          handleClose={[Function]}
          prefixCls="ui-alert"
        >
          <Icon
            className="ui-alert-close-btn"
            onClick={[Function]}
            prefixCls="zw-icon"
            size="md"
            theme="default"
            type="wrong"
            viewBox="0 0 32 32"
          >
            <Iconfont
              className="zw-icon ui-alert-close-btn zw-icon--wrong zw-icon--default zw-icon--md"
              onClick={[Function]}
              size="md"
              theme="default"
              type="wrong"
            >
              <Icon
                className="zw-icon ui-alert-close-btn zw-icon--wrong zw-icon--default zw-icon--md"
                onClick={[Function]}
                prefixCls="zw-icon"
                size="md"
                theme="default"
                type="wrong"
                viewBox="0 0 32 32"
              >
                <i
                  className="zw-icon ui-alert-close-btn zw-icon--wrong zw-icon--default zw-icon--md"
                  onClick={[Function]}
                  size="md"
                  theme="default"
                  type="wrong"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    viewBox="0 0 32 32"
                    width="1em"
                  >
                    <use
                      xlinkHref="#wrong"
                    />
                  </svg>
                </i>
              </Icon>
            </Iconfont>
          </Icon>
        </CloseIcon>
        <span
          className="ui-alert-message"
        >
          基础用法-信息提示
        </span>
        <span
          className="ui-alert-description"
        >
          可关闭提示框
        </span>
      </div>
    </Alert>
  </LocaleReceiver>
</ForwardRef(forwardRef)>
`;
